File: python/src/traffic_control/network/graph.py
from __future__ import annotations

import networkx as nx
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from traffic_control.models import Network, Junction, Road


def to_networkx(network: "Network") -> nx.DiGraph:
    G = nx.DiGraph()
    for junction in network.junctions:
        G.add_node(junction.id, position=junction.position, external_flow=junction.external_flow)
    for road in network.roads:
        G.add_edge(
            road.source,
            road.target,
            id=road.id,
            capacity=road.capacity,
            length=road.length,
            cost=road.cost_per_unit,
            flow=road.flow,
        )
    return G


def validate_flow_conservation(network: "Network", flows: dict[str, float] | None = None) -> tuple[bool, list[str]]:
    violations = []
    if flows is None:
        flows = {r.id: r.flow for r in network.roads}

    for junction in network.junctions:
        inflow = sum(flows.get(r.id, 0) for r in network.roads if r.target == junction.id)
        outflow = sum(flows.get(r.id, 0) for r in network.roads if r.source == junction.id)
        net = inflow - outflow
        if abs(net - junction.external_flow) > 1e-6:
            violations.append(f"Junction {junction.id}: net flow {net:.2f} != external {junction.external_flow}")

    return len(violations) == 0, violations


def validate_capacities(network: "Network", flows: dict[str, float] | None = None) -> tuple[bool, list[str]]:
    violations = []
    if flows is None:
        flows = {r.id: r.flow for r in network.roads}

    for road in network.roads:
        flow = flows.get(road.id, 0)
        if flow < -1e-6:
            violations.append(f"Road {road.id}: negative flow {flow:.2f}")
        if flow > road.capacity + 1e-6:
            violations.append(f"Road {road.id}: flow {flow:.2f} exceeds capacity {road.capacity}")

    return len(violations) == 0, violations


def validate_connectivity(network: "Network") -> tuple[bool, list[str]]:
    violations = []
    G = to_networkx(network)

    if not nx.is_weakly_connected(G):
        violations.append("Network is not weakly connected")

    sources = [j.id for j in network.junctions if j.external_flow > 0]
    sinks = [j.id for j in network.junctions if j.external_flow < 0]

    if not sources:
        violations.append("No source junctions (external_flow > 0)")
    if not sinks:
        violations.append("No sink junctions (external_flow < 0)")

    for source in sources:
        for sink in sinks:
            if not nx.has_path(G, source, sink):
                violations.append(f"No path from source {source} to sink {sink}")

    return len(violations) == 0, violations


def validate_network(network: "Network", flows: dict[str, float] | None = None) -> tuple[bool, list[str]]:
    all_violations = []

    ok, v = validate_flow_conservation(network, flows)
    all_violations.extend(v)

    ok, v = validate_capacities(network, flows)
    all_violations.extend(v)

    ok, v = validate_connectivity(network)
    all_violations.extend(v)

    return len(all_violations) == 0, all_violations


def find_bottlenecks(network: "Network", flows: dict[str, float] | None = None, threshold: float = 0.8) -> list[tuple[str, float]]:
    if flows is None:
        flows = {r.id: r.flow for r in network.roads}

    bottlenecks = []
    for road in network.roads:
        flow = flows.get(road.id, 0)
        utilization = flow / road.capacity if road.capacity > 0 else 0
        if utilization >= threshold:
            bottlenecks.append((road.id, utilization))

    return sorted(bottlenecks, key=lambda x: x[1], reverse=True)


def compute_shortest_paths(network: "Network", weight: str = "length") -> dict[str, dict[str, float]]:
    G = to_networkx(network)
    return dict(nx.all_pairs_dijkstra_path_length(G, weight=weight))


def compute_betweenness_centrality(network: "Network") -> dict[str, float]:
    G = to_networkx(network)
    return nx.betweenness_centrality(G, weight="length")

File: rust/src/network/graph.rs
use petgraph::graph::{DiGraph, NodeIndex};
use petgraph::visit::EdgeRef;
use std::collections::HashMap;
use crate::models::{Junction, Network, Road};

pub fn to_petgraph(network: &Network) -> DiGraph<Junction, Road> {
    let mut graph = DiGraph::new();
    let mut node_map = HashMap::new();

    for junction in &network.junctions {
        let idx = graph.add_node(junction.clone());
        node_map.insert(junction.id.clone(), idx);
    }

    for road in &network.roads {
        if let (Some(&src), Some(&tgt)) = (node_map.get(&road.source), node_map.get(&road.target)) {
            graph.add_edge(src, tgt, road.clone());
        }
    }

    graph
}

pub fn validate_flow_conservation(network: &Network, flows: &HashMap<String, f64>) -> Result<(), Vec<String>> {
    let mut violations = Vec::new();

    for junction in &network.junctions {
        let inflow: f64 = network.roads.iter()
            .filter(|r| r.target == junction.id)
            .map(|r| flows.get(&r.id).copied().unwrap_or(0.0))
            .sum();

        let outflow: f64 = network.roads.iter()
            .filter(|r| r.source == junction.id)
            .map(|r| flows.get(&r.id).copied().unwrap_or(0.0))
            .sum();

        let net_flow = inflow - outflow;
        if (net_flow - junction.external_flow).abs() > 1e-6 {
            violations.push(format!(
                "Junction {}: inflow - outflow = {:.2f}, expected {:.2f}",
                junction.id, net_flow, junction.external_flow
            ));
        }
    }

    if violations.is_empty() {
        Ok(())
    } else {
        Err(violations)
    }
}

pub fn validate_capacities(network: &Network, flows: &HashMap<String, f64>) -> Result<(), Vec<String>> {
    let mut violations = Vec::new();

    for road in &network.roads {
        let flow = flows.get(&road.id).copied().unwrap_or(0.0);
        if flow < -1e-6 {
            violations.push(format!("Road {}: negative flow {:.2f}", road.id, flow));
        }
        if flow > road.capacity + 1e-6 {
            violations.push(format!(
                "Road {}: flow {:.2f} exceeds capacity {:.2f}",
                road.id, flow, road.capacity
            ));
        }
    }

    if violations.is_empty() {
        Ok(())
    } else {
        Err(violations)
    }
}

pub fn validate_connectivity(network: &Network) -> Result<(), Vec<String>> {
    let graph = to_petgraph(network);
    let mut violations = Vec::new();

    if !petgraph::algo::is_connected(graph.as_undirected()) {
        violations.push("Network is not weakly connected".to_string());
    }

    let sources: Vec<_> = network.junctions.iter()
        .filter(|j| j.external_flow > 0.0)
        .collect();
    let sinks: Vec<_> = network.junctions.iter()
        .filter(|j| j.external_flow < 0.0)
        .collect();

    if sources.is_empty() {
        violations.push("No source junctions (external_flow > 0)".to_string());
    }
    if sinks.is_empty() {
        violations.push("No sink junctions (external_flow < 0)".to_string());
    }

    for source in sources {
        for sink in sinks {
            let src_idx = graph.node_indices().find(|n| graph[*n].id == source.id);
            let tgt_idx = graph.node_indices().find(|n| graph[*n].id == sink.id);

            if let (Some(src), Some(tgt)) = (src_idx, tgt_idx) {
                if !petgraph::algo::has_path_connecting(&graph, src, tgt, None) {
                    violations.push(format!(
                        "No path from source {} to sink {}",
                        source.id, sink.id
                    ));
                }
            }
        }
    }

    if violations.is_empty() {
        Ok(())
    } else {
        Err(violations)
    }
}

pub fn validate_network(network: &Network, flows: &HashMap<String, f64>) -> Result<(), Vec<String>> {
    let mut all_violations = Vec::new();

    if let Err(v) = validate_flow_conservation(network, flows) {
        all_violations.extend(v);
    }
    if let Err(v) = validate_capacities(network, flows) {
        all_violations.extend(v);
    }
    if let Err(v) = validate_connectivity(network) {
        all_violations.extend(v);
    }

    if all_violations.is_empty() {
        Ok(())
    } else {
        Err(all_violations)
    }
}